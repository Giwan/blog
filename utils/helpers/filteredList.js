/**
 * Filters a list of items based on the array of filter items
 * Each tool has a list of labels. 
 * The filter list is a list of labels that should match.
 * @param {Array} tools 
 * @param {Array} filters 
 * @returns 
 */
 export const filteredList = (tools, filters) => {
    if (!filters?.length) {
        return tools;
    }

    return tools.filter((t) => filters.find((f) => t.labels?.includes(f)));
};

export default filteredList;
