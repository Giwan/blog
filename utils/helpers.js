import filteredList from "./helpers/filteredList";

export { filteredList };


export const reverseDate = (date = "") =>
    parseInt(date.split("-").reverse().join(""));

/**
 * Format the string date to a number
 * Typically used for sorting by date
 * @param {String} date A date string that formatted as 10-10-1980
 * @returns Number
 */
export const getDateNumber = (dateString) => {
    if (typeof dateString !== "string") {
        throw new Error("Provided date argument is not of type string");
    }
    return Number(dateString?.replace(/-/g, "")) || 0;
}

/**
 * Check to see what style should be applied.
 * This is used by the navigation route.
 * It is in this file mostly because of testing.
 * @param {Object} router The router parameters object
 * @param {Object} styles The imported styles object
 * @param {Object | String} target The target path
 */
export const getStyle = (router, styles, target) => {
    const [_target, _routes] = getStyleValidation(router, target);

    // early check for exact match
    if (router.pathname === _target) {
        return styles.activeLink;
    }

    if (Array.isArray(_routes) && _routes.length) {
        return _routes.find((route) => router.pathname.indexOf(route) > -1)
            ? styles.activeLink
            : undefined;
    }
};

/**
 * Throws an error if any important information is missing
 * when determining the routes.
 * @param {Object} router
 * @param {String | Object} target
 */
export const getStyleValidation = (router, target) => {
    if (!(router && router.pathname)) {
        throw new Error("Please provide a valid router object with pathname");
    }

    let _target = target,
        _routes = [];
    if (typeof target === "object") {
        if (!target.path) {
            throw new Error(
                "The path value is required when the target is an object. target: " +
                    JSON.stringify(target)
            );
        }
        _target = target.path;
        _routes = target.routes;
    }

    return [_target, _routes];
};
