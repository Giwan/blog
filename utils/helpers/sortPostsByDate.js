import { getDateNumber } from "../helpers";

/**
 * Sort posts by published date
 * @param {*} p1
 * @param {*} p2
 * @returns
 */
export const sortPostsByDate = (dateName) => (p1, p2) => {
    if (!(typeof dateName === "string" && dateName.length > 0)) {
        throw new Error("A valid dateName value is required");
    }

    const [date1, date2] = [p1, p2].map((date) => {
        const meta = date?.module?.meta;

        const dateString = meta[dateName];
        if (typeof dateString !== "string") {
            throw new Error(
                `Invalid ${dateName} date. Value of type string expected. Received: ${dateString}`
            );
        }
        return getDateNumber(dateString);
    });

    return date2 - date1;
};
