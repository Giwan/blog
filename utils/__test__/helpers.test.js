import { getDateNumber } from "../helpers";

describe("get date number", () => {
    it("should return a date number", () => {
        expect(() => getDateNumber()).toThrowErrorMatchingSnapshot();
        expect(() => getDateNumber(1)).toThrowErrorMatchingSnapshot();
        expect(getDateNumber("")).toBe(0);
        expect(getDateNumber("-")).toBe(0);
        expect(getDateNumber("somerandomstring")).toBe(0);
        expect(getDateNumber("10-10-1980")).toBe(10101980);
    });
});
