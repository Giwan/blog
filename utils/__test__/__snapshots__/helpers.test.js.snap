// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get date number should return a date number 1`] = `"Provided date argument is not of type string"`;

exports[`get date number should return a date number 2`] = `"Provided date argument is not of type string"`;
